### PR TITLE
[TFLite int16] 16-bit support for reference kernels: MAX/MIN and PACK/UNPACK

### PR DESCRIPTION
**Pull request by wwwind**
This PR is one of steps to extend 8-bit quantization to support symmetric 16-bit activations.

Each activation is of type int16 and symmetric around zero. The weight tensor precision remains at 8-bit signed values. The bias is set to int64 precision.

In this PR we introduce implementation and tests for MAXIMUM/MINIMUM kernel reference functions and for PACK/UNPACK kernel reference functions.
The specification of these operators:

MAXIMUM 
  Input 0: 
    data_type  : int16 
    range      : [-32768, 32767] 
    granularity: per-tensor, zero_point=0 
  Output 0: 
    data_type  : int16 
    range      : [-32768, 32767] 
    granularity: per-tensor, zero_point=0 
    restriction: Input and outputs must all have same scale 
 
MINIMUM 
  Input 0: 
    data_type  : int16 
    range      : [-32768, 32767] 
    granularity: per-tensor, zero_point=0 
  Output 0: 
    data_type  : int16 
    range      : [-32768, 32767] 
    granularity: per-tensor, zero_point=0 
    restriction: Input and outputs must all have same scale 

UNPACK 
  Input 0: 
    data_type  : int16 
    range      : [-32768, 32767] 
    granularity: per-tensor, zero_point=0 
  Outputs 0-N: 
    data_type  : int16 
    range      : [-32768, 32767] 
    granularity: per-tensor, zero_point=0 
    restriction: Input and outputs must all have same scale

### CODE REVIEW

**Comment by wwwind** [comment]
Hi @suharshs , Could you please review this PR with a set of 16-bit reference kernels ? Thanks!

**Review by suharshs — DISMISSED** [review]

**Comment by gbaned** [comment]
Here are internal failures , can you please check.

third_party/tensorflow/lite/kernels/test_util.cc:329
Expected equality of these values:
  CountPartitionsDelegatedTo(interpreter_.get(), TestNnApiDelegate())
    Which is: 0
  1
Expecting operation to be accelerated but cannot find a partition associated to the NNAPI delegate
third_party/tensorflow/lite/kernels/test_util.cc:329
Expected equality of these values:
  CountPartitionsDelegatedTo(interpreter_.get(), TestNnApiDelegate())
    Which is: 0
  1
Expecting operation to be accelerated but cannot find a partition associated to the NNAPI delegate

**Comment by jdduke** [comment]
Many of our delegates key off of the operator version to know whether it's supported. Are we updating the op versioning for these new 16-bit kernels?

**Comment by wwwind** [comment]
@jdduke No, I don't do this.

**Comment by wwwind** [comment]
@jdduke About the versioning: 
We introduced a new input/output type for operators. Do we need to update version for them and change the function GetMinimumRuntimeVersionForModel ?
We could change GetBuiltinOperatorVersion() function to return a special version that indicates that our input/output is 16-bit. Does it make sense ?

I am not sure how to proceed here. I really appreciate your help what to do with these failures.
Thanks!

**Comment by jdduke** [comment]
Right, having the updated op version should help resolve some of the internal test failures we're seeing.

To do this, you can add [logic in GetBuiltinOperatorVersion to detect the input/output types](https://github.com/tensorflow/tensorflow/blob/306dee4096e97520b1218d78c23b1a72941a9508/tensorflow/lite/tools/versioning/op_version.cc), and use an incremented version, update the op version to TFlite version mapping [here](https://github.com/tensorflow/tensorflow/blob/306dee4096e97520b1218d78c23b1a72941a9508/tensorflow/lite/toco/tflite/op_version.cc), and then update the max version where the builtin is registered [here](https://github.com/tensorflow/tensorflow/blob/306dee4096e97520b1218d78c23b1a72941a9508/tensorflow/lite/kernels/register.cc#L43). 

I realize this process is fairly tedious, we're hoping to remove some of the boilerplate, but for now that's the general pattern.

**Comment by wwwind** [comment]
@jdduke Hi Jared ! Thanks. I updated as you suggested.
I have a couple of quesions:
1. Can I reproduce tests that failed in my environment ? What is a bazel task to run these tests ? Do I need pull external dependencies ?
2. We have a set of 16x8 kernels merged without versioning. Do we need to update versioning for all 16x8 kernels that have been merged or under review ?
3. Do we need to add tests for versioning ? I see [op_version_test](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/versioning/op_version_test.cc), for example. 
4. General question: Is this versioning applied when we change syntax of the operator, algorithm inside ?
Thanks !

**Comment by wwwind** [comment]
Hi @jdduke, Could you please review this PR ?

File: tensorflow/lite/tools/versioning/op_version.cc [hunk 1]
```diff
@@ -287,15 +291,26 @@ int GetBuiltinOperatorVersion(const OpSignature& op_sig) {
       }
       return 1;
 
+    case BuiltinOperator_MAXIMUM:
+    case BuiltinOperator_MINIMUM:
+    case BuiltinOperator_PACK:
+      if (op_sig.input_types.at(0) == TensorType_INT8) {
```
Comment (jdduke) [inline tensorflow/lite/tools/versioning/op_version.cc]: I would avoid grouping pack w/ MAX/MIN, OK to have 2 separate case branches.

**Review by jdduke — CHANGES_REQUESTED** [review]

File: tensorflow/lite/tools/versioning/op_version.cc (hunk 1 above)
Comment (wwwind) [reply tensorflow/lite/tools/versioning/op_version.cc]: @jdduke done

**Review by jdduke — DISMISSED** [review]

**Comment by wwwind** [comment]
Hi @jdduke, Could you please re-approve this PR ? There was a merge conflict and I had to resolve it. Thanks!

**Review by jdduke — APPROVED** [review]

**Comment by wwwind** [comment]
Hi @rthadur , Sorry to bother you, but it looks that this PR has stuck as well. 
Is it something internally ? 
Thanks!

**Review by mihaimaruseac — APPROVED** [review]

**Comment by OriAlpha** [comment]
On which version of TensorFlow is this working??

**Pull request merged**